bin/wasm-node/rust/src/platform.rs

```diff
@@ -106,183 +104,463 @@ impl smoldot_light_base::Platform for Platform {
+            ConnectionInner::NotOpen => unreachable!(),
+            ConnectionInner::SingleStream => {
+                unsafe {
+                    bindings::connection_close(self.0 .0);
+                }
+
+                debug_assert_eq!(self.0 .1, 0);
+                true
+            }
+            ConnectionInner::MultiStream {
+                connection_handles_alive,
+                ..
+            } => {
+                unsafe { bindings::connection_stream_close(self.0 .0, self.0 .1) }
+                *connection_handles_alive -= 1;
+                let remove_connection = *connection_handles_alive == 0;
+                if remove_connection {
+                    unsafe {
+                        bindings::connection_close(self.0 .0);
+                    }
+                }
+                remove_connection
+            }
+            ConnectionInner::Closed {
+                connection_handles_alive,
+                ..
+            } => {
+                *connection_handles_alive -= 1;
+                let remove_connection = *connection_handles_alive == 0;
+                if remove_connection {
+                    unsafe {
+                        bindings::connection_close(self.0 .0);
+                    }
+                }
+                remove_connection
+            }
+        };
+
+        lock.streams.remove(&(self.0 .0, self.0 .1)).unwrap();
 
-impl fmt::Debug for Connection {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        f.debug_tuple("Connection")
-            .field(self.id.as_ref().unwrap())
-            .finish()
+        if remove_connection {
+            lock.connections.remove(&self.0 .0).unwrap();
+        }
     }
 }
 
-impl Drop for Connection {
+pub(crate) struct ConnectionWrapper(u32);
+
+impl Drop for ConnectionWrapper {
     fn drop(&mut self) {
-        if let Some(id) = self.id {
-            unsafe {
-                bindings::connection_close(id);
+        let mut lock = STATE.try_lock().unwrap();
+
+        let connection = lock.connections.get_mut(&self.0).unwrap();
+        let remove_connection = match &mut connection.inner {
+            ConnectionInner::NotOpen | ConnectionInner::SingleStream => unreachable!(),
+            ConnectionInner::MultiStream {
+                connection_handles_alive,
+                ..
+            }
+            | ConnectionInner::Closed {
+                connection_handles_alive,
+                ..
+            } => {
+                *connection_handles_alive -= 1;
+                *connection_handles_alive == 0
+            }
+        };
+
+        if remove_connection {
+            lock.connections.remove(&self.0).unwrap();
+            if remove_connection {
+                unsafe {
+                    bindings::connection_close(self.0);
+                }
             }
         }
     }
 }
 
+lazy_static::lazy_static! {
+    static ref STATE: Mutex<NetworkState> = Mutex::new(NetworkState {
```

Comment:
is it a common / preferred way of writing "bindings" code? ie having global state as a const? just curious

Reply:
Well, this code is specifically to bind the light client to the "wasm platform" capabilities, so I went for a global const because the "platform" is a global thing.
A similar example is that on Linux the list of open file descriptors is a global "const" as well.